src/match.c

```diff
@@ -1,128 +1,219 @@
+    }
+
+    match->word = strdup(word);
+    match->lineNum = lineNum;
+    match->position = position;
+    match->line = strdup(line);
 
-    match->word = word;
-    match->line = line;
-    
     return 0;
 }
 
-int free_match(match* match)
-{
-    if(match == NULL) return -1;
-    free(match);
+int match_set_line(match* match, char* line) {
+    if (match == NULL) {
+        return -1;
+    }
+
+    match->line = strdup(line);
     return 0;
 }
 
-char* get_word(match* match)
-{
-    if(match == NULL) return NULL;
+int match_free(match* match) {
+    if (match == NULL) {
+        return -1;
+    }
+
+    if (match->word != NULL) {
+        free(match->word);
+    }
+
+    if (match->line != NULL) {
+        free(match->line);
+    }
+
+    free(match);
+
+    return 0;
+}
+char* match_get_word(match* match) {
+    if (match == NULL) {
+        return NULL;
+    }
     return match->word;
 }
 
-int get_line(match* match)
-{
-    if(match == NULL) return -1;
+char* match_get_line(match* match) {
+    if (match == NULL) {
+        return NULL;
+    }
     return match->line;
 }
 
-//changed inputs
-match* next_match(match* match, GList* matches)
-{
-    if (match == NULL) return NULL;
-    GList* cur = g_list_find(matches, match);
-    if (cur->next == NULL) return (g_list_first(matches))->data;
-    return (cur->next)->data;
-}
-
-//changed inputs
-match* prev_match(match* match, GList* matches)
-{
-    if (match == NULL) return NULL;
-    GList* cur = g_list_find(matches, match);
-    if (cur->prev == NULL) return (g_list_last(matches))->data;
-    return (cur->prev)->data;
-}
-//changed inputs, return type
-GList* insert_at(match* newMatch, int index, GList* matches)
-{
-    matches = g_list_insert(matches, newMatch, index);
-    return matches;
-}
-
-//changed inputs, return type
-GList* append_(match* newMatch, GList* matches)
-{
-    matches = g_list_append(matches, newMatch);
-    return matches;
-}
-
-//changed inputs, return type
-GList* remove_at(int index, GList* matches)
-{
-    //get element at position
-    GList *toDelete = g_list_nth(matches, index);
-    
-    //remove position at element
-    matches = g_list_delete_link(matches, toDelete);
-    return matches;
-}
-
-//changed inputs
-match* get_at_index(int index, GList* matches)
-{
-    if (g_list_length(matches) == 0) return NULL;
-    GList *toReturn = g_list_nth(matches, index);
-    if (toReturn == NULL) return NULL;
-    return toReturn -> data;
-}
-
-//changed inputs
-int get_index(match* match, GList* matches)
-{
-    int position = g_list_index(matches, match);
+int match_get_line_num(match* match) {
+    if (match == NULL) {
+        return -1;
+    }
+    return match->lineNum;
+}
+
+int match_get_position(match* match) {
+    if (match == NULL) {
+        return -1;
+    }
+    return match->position;
+}
+
+
+match* match_next(match* match, list_t* matches) {
+    if (match == NULL) {
+        return NULL;
+    }
+    int rc = list_locate(matches, match);
```

Comment:
This is an O(n) operation. Can we make it O(1)? Linked list next and prev should be O(1).